lib/puppet/util/windows/service.rb

```diff
@@ -562,12 +568,13 @@ def open_scm(scm_access, &block)
       def transition_service_state(service_name, valid_initial_states, final_state, &block)
         service_access = SERVICE_START | SERVICE_STOP | SERVICE_PAUSE_CONTINUE | SERVICE_QUERY_STATUS
         open_service(service_name, SC_MANAGER_CONNECT, service_access) do |service|
-          status = query_status(service)
-          initial_state = status[:dwCurrentState]
-
+          initial_state = nil
+          query_status(service) do |status|
+            initial_state = status[:dwCurrentState]
```

Comment:
In Ruby, `=` assigns the same object to a variable, see for example
```
"2.4.2 :001 > a = "foo"
 => "foo"
2.4.2 :002 > b = a
 => "foo"
2.4.2 :003 > b.object_id
 => 70151442240220
2.4.2 :004 > a.object_id
 => 70151442240220
```

meaning checking the `initial_state` afterwards could be a use-after-free scenario. Maybe worth checking if that is the case here? If so, there's other places in the code that may run into this issue.

Reply:
I left a a comment about this in the code, I investigated this scenario and FFI does what we want it to in this scenario: during the assignment of an integer value stored in memory FFI will create a new ruby `Integer ` object in ruby memory to represent the value that was read from memory. That newly created ruby `Integer` is entirely pure ruby at that point, and safe to use outside the context of struct even after the memory is freed.

Reply:
Scratch that, I didn't have to do this after all.

See general comment below